feat(parser) Integrate a Simple Arithmetic Expression Parser

ayirr7 opened the pull request:
The simple arithmetic expression parser is in simple_arithm_expression.py. This is currently able to parse arithmetic expressions involving "+" and "*" (the addition and multiplication operators). As of now, it does not support the "-" and "/" (the subtraction and division operators), parenthetical expressions and grouping, or other accessories, like user-inputted spaces/tab characters. These would probably be some of the next steps. This will be handled most likely by writing a longer and slightly more complex grammar, with their own visitor functions. 

Another potential limitation of the current implementation is that there is a check for "None." This came about from the way an empty space has been handled. In visit_space, a space is reached and nothing is returned. This allows us to discard '' and build a Literal() with only the needed operand (an integer). If there is a cleaner/smarter approach for handling this empty space, that would be a good improvement.

ayirr7 committed:
Draft arithmetic expressions parser.

ayirr7 committed:
Merge branch 'master' into feat/arithmeticExpParser

ayirr7 committed:
Add an expression parser which produces AST representations of simple arithmetic expressions (+, *).

ayirr7 committed:
Format fix for simple arithmetic expression parser.

fpacifici commented on snuba/query/parser/arithmetic_expressions.py:
```diff
@@ -0,0 +1,50 @@
+from parsimonious.grammar import Grammar
```
you do not need this file

ayirr7 replied:
Removed.

fpacifici commented on snuba/query/parser/simple_arithm_expressions.py:
```diff
@@ -0,0 +1,50 @@
+            # that arose from empty space ''
+            return factor
+        else:
+            return FunctionCall(None, term[0][0], (factor, term[0][1]))
```
We have some helper functions to build the right FunctionCall structure (that uses the right function name for arithmetic operations)
https://github.com/getsentry/snuba/blob/master/snuba/query/dsl.py#L25

ayirr7 replied:
New version uses them.

fpacifici reviewed (commented):
Conceptually seems correct, now please let's add it to the grammar here 
https://github.com/getsentry/snuba/blob/master/snuba/query/parser/expressions.py#L19

So we can actually test it.

ayirr7 committed:
First step of integrating arithmetic expression parsing.

ayirr7 committed:
Integrated "Simple Arithmetic Expression Parser" into existing parser.

ayirr7 committed:
Fixed Simple Arithmetic Expression Parser (generic_visit is identical to existing parser's now).

ayirr7 committed:
Merge branch 'master' into feat/arithmeticExpParser

ayirr7 commented:
Simple Arithmetic Expression parser (described above) was integrated into existing expression parser. The missing features noted above will be next steps. Furthermore, this integration only supports multiplication of numeric_literals, but column_name and function_call should also be supported. Grammar rule names and rules themselves will probably be also modified.

fpacifici commented on snuba/query/parser/simple_arithm_expressions.py:
```diff
@@ -0,0 +1,59 @@
+from parsimonious.grammar import Grammar
```
you do not need this file. Please remove it.

ayirr7 replied:
Removed.

fpacifici reviewed (commented):
Please add some unit tests to tests/query/parser/test_expressions.py to test your code.

ayirr7 committed:
Deleted standalone version of Arithmetic Expression Parser.

lynnagara reviewed (commented):
Thanks, this is a nice enhancement. A few questions about which cases we plan to support here:

Currently this looks like it handles the case where multiply happens before add. So `5*5+1` is allowed but not `5+1*5`. Is that right? Why is that decision made? Related to this, I think it'd be really useful if you can add some tests so we can get an idea of which cases we want to support and which not.

Also do we want to handle expression with parens like `(5+1)*5` and apply those operations in the right order?

I think we probably also want to handle/ignore any spaces passed. e.g. `5+1` and `5 + 1` and `5+  1` are all pretty much the same thing.

ayirr7 committed:
Add unit test cases to test integration of arithmetic expression parser.

ayirr7 committed:
Update to arithmetic expression parser - top-level parentheses (in root element), space handling, renaming and reorganization.

ayirr7 committed:
Merge branch 'master' into feat/arithmeticExpParser

ayirr7 commented:
Updates to integrated arithmetic expression parser: 

- Parentheses were added to the top-level element of an arithmetic expression (i.e. in root_element). Currently this is being handled by the creation of a new grammar rule by the name of "top_arithm_expression."
- Space handling so that "1   * 5+2", "1*5 +      2" and all variations are equivalent.
- Renaming of "simple_term2" and "arithmetic_expression" by assigning "high_pri_arithmetic" and "low_pri_arithmetic" names to high priority arithmetic expressions (those with "*" operator) and low priority arithmetic expressions (those with "+" operator), respectively.
- Deletion of unnecessary file(s).

Next steps:

- Modification of grammar to include subtraction and division. (This will probably eliminate recursive elements)
- Support parenthetical expressions to indicate priority (i.e. 2+5 * (4-3) )

ayirr7 committed:
Deleted arithmetic_expressions.py

ayirr7 commented on snuba/query/parser/expressions.py:
```diff
@@ -15,28 +15,33 @@
+parameter             = expression space* comma space*
+function_call         = function_name open_paren parameters_list? close_paren (open_paren parameters_list? close_paren)?
+simple_term           = quoted_literal / numeric_literal / column_name
+simple_term2          = numeric_literal (("*" simple_term2) / empty)
```
I renamed the grammar rules - if you get a chance to see them, let me know what you think!

ayirr7 commented on snuba/query/parser/expressions.py:
```diff
@@ -15,28 +15,33 @@
-comma           = ","
+root_element          = function_call / function_name
+expression            = function_call / arithmetic_expression / simple_term
+arithmetic_expression = simple_term2 (("+" arithmetic_expression) / empty)
```
No, the second part isn't optional. The point of "empty" is so that there can be matching with either a) another arithmetic expressions or b) nothing at all. So the "empty" is necessary, as far as I can tell. 

Yes, the current version intends for arithmetic_expression to refer to itself (recursive behavior). In future versions, there is a good chance that will be replaced by regex repetition via the "+" you mentioned.

ayirr7 replied:
Finally, the spacing issue you mentioned was handled in the latest update. Both 5*5+1 and 5+1*5 are handled. It does not matter which order the multiplication and/or addition operators appear in the expression. 

Order of operations with parentheses (for e.g. 3+4*(2-1) ) is a future step.

ayirr7 committed:
Modified one test case to address a potential concern (order of operators).

fpacifici commented on snuba/query/parser/expressions.py:
```diff
@@ -15,28 +15,34 @@
-space           = " "
-comma           = ","
+root_element          = function_call / function_name / top_arithm_expression
+expression            = function_call / simple_term
```
It seems that, this way, arithmetic expression can only be directly at the root_level and they cannot be nested into functions.
This should be valid:
```
f(a * b, g(c * 3))
```
I think you should have low_pri_arithmetic also somewhere on the right side of this rule.

ayirr7 replied:
Added low_pri_arithmetic to expression grammar rule.

fpacifici commented on snuba/query/parser/expressions.py:
```diff
@@ -52,6 +58,32 @@ def visit_function_name(self, node: Node, visited_children: Iterable[Any]) -> st
     def visit_column_name(self, node: Node, visited_children: Iterable[Any]) -> Column:
         return Column(None, None, node.text)
 
+    def visit_empty(self, node, children):
```
Please provide type hints and return value for all these methods.

ayirr7 replied:
Added.

fpacifici commented on snuba/query/parser/expressions.py:
```diff
@@ -52,6 +58,32 @@ def visit_function_name(self, node: Node, visited_children: Iterable[Any]) -> st
         return Column(None, None, node.text)
 
+    def visit_empty(self, node, children):
+        return
```
if you need to return None, please explicitly do so
```
return None
```

fpacifici commented on snuba/query/parser/expressions.py:
```diff
@@ -150,7 +182,9 @@ def parse_aggregation(
     used in production).
     """
     expression_tree = minimal_clickhouse_grammar.parse(aggregation_function)
+    print(expression_tree)
```
please remove the print statements

ayirr7 replied:
Removed.

fpacifici commented on tests/query/parser/test_expressions.py:
```diff
@@ -9,6 +9,81 @@
+            ),
+        ),
+    ),  # Combination of multiplication, addition
+    (
+        ["(3*2+4*5  )", None, None],
+        FunctionCall(
+            None,
+            "plus",
+            (
+                FunctionCall(None, "multiply", (Literal(None, 3), Literal(None, 2))),
+                FunctionCall(None, "multiply", (Literal(None, 4), Literal(None, 5))),
+            ),
+        ),
+    ),  # Symmetric Combination of multiplication, addition
```
hint: you can remove this test since the following one is testing the same cases.

ayirr7 replied:
Removed.

fpacifici reviewed (changes requested):
This is good, thanks. 
Please see a few comments inline to support the missing cases as well.

ayirr7 commented on snuba/query/parser/expressions.py:
```diff
@@ -15,28 +15,34 @@
+expression            = function_call / simple_term
+top_arithm_expression = open_paren low_pri_arithmetic close_paren
+low_pri_arithmetic    = space* high_pri_arithmetic space* (("+" low_pri_arithmetic) / empty)
+high_pri_arithmetic   = space* (numeric_literal / column_name) space* (("*" high_pri_arithmetic) / empty)
```
Added function_call to high_pri_arithmetic grammar rule.

ayirr7 committed:
Added few test cases (arithm expressions with function_call, function_call with a nested arithm expression).

ayirr7 committed:
Modified grammar rules to account for function calls with nested arithmetic expressions, and arithmetic expressions consisting of function calls. Added type hints and return values, along with other cleanups.

ayirr7 committed:
Merge branch 'master' into feat/arithmeticExpParser

fpacifici commented on snuba/query/parser/expressions.py:
```diff
@@ -52,6 +56,38 @@ def visit_function_name(self, node: Node, visited_children: Iterable[Any]) -> st
     def visit_column_name(self, node: Node, visited_children: Iterable[Any]) -> Column:
         return Column(None, None, node.text)
 
+    def visit_empty(self, node: Node, visited_children: Iterable[Any]) -> None:
```
`-> None` means this does not return anything (like void), which yeah in python kind of means the same but for the type checker it is not the same thing. 
Please return `Optional[Expression]` which allows you to return None.
Also could you please run `mypy --strict snuba/query/parser/expressions.py` I suspect this would not type check even if it behaves properly.

ayirr7 replied:
Type checking done.

fpacifici commented on tests/query/parser/test_expressions.py:
```diff
@@ -9,6 +9,113 @@
 
 test_data = [
+    (
+        ["f(a *   b, g(c * 3))", None, None],
```
one minor nit on your test cases. For unit tests it is advisable to have specific tests that test individual interesting inputs and then, if needed, some more complex tests that put things together. This helps us getting a better sense of the logical coverage (how well the test cover the interesting cases) we have quickly.
In this case I would advise you to have one or two specific test to stress the spaces in the expressions, and just rely on a standard spacing pattern for the others.

ayirr7 replied:
Good point. Removed some cases of extra spacing.

fpacifici reviewed (commented):
This is excellent, thanks , there are a couple of minor comments inline specifically about the typing, otherwise it is good.
@lynnagara  is it good to you ?

fpacifici commented:
Also could you please do a sanity test by running the sentry tests against snuba?
ensure you pull Sentry master, then the procedure is explained here
It is explained here: https://github.com/getsentry/snuba#testing-against-sentry

lynnagara reviewed (commented):
Thanks, I think this is looking good.

I'm not sure how likely this is but I think it'll be worth also adding a test for a string that looks like an arithmetic expression. Like if a date "2020-06-06" gets passed as an expression.

ayirr7 committed:
Remove excessive spacing in some test cases.

ayirr7 committed:
Revised type hinting in visitor methods.

ayirr7 commented on snuba/query/parser/expressions.py:
```diff
@@ -21,22 +21,26 @@
-comma           = ","
+root_element          = function_call / function_name / top_arithm_expression
+expression            = low_pri_arithmetic / function_call / simple_term
+top_arithm_expression = open_paren low_pri_arithmetic close_paren
```
For the first point - testing a string that looks like an arithmetic expression. I think this is a great idea, and I will definitely try to integrate it asap. Currently though, in this PR, the parser does not support subtraction or division. As per my knowledge, dates can be represented with the "-" or "/" signs, so perhaps once the next round of modifications to the grammar are applied, I can add this as a test case.

Second point - currently, a top-level arithmetic expression needs to be framed by a set of parentheses. This means that if the root_element is an arithmetic expression itself, it will be in parentheses (for example, (2 * 3 + 4) ). This does not mean that each unit of arithmetic ( 2 * 3 or 3 + 4 etc) are in parentheses. The top-level parentheses were added in order to avoid confusion in cases like "a + b," where "a" and "b" could be mistakenly resolved as function_name(s). So, to distinguish the fact that something like "a + b" should be recognized as an arithmetic expression, we decided that top-level arithmetic expressions need a differentiating factor - the parentheses. Hope this was a clear explanation!

fpacifici replied:
> For the first point - testing a string that looks like an arithmetic expression. I think this is a great idea, and I will definitely try to integrate it asap

I don't think this is needed. String literal are always quoted. So, in an expression `2020-06-06` would be basically invalid (or at best be an arithmetic expression). To pass a string literal that represents a date you do this `'2020-06-06'`.

lynnagara replied:
Yeah. I am talking about quoted strings. The reason I mentioned having a test for this is that I noticed arithmetic expressions takes priority over quoted literals now (line 25).

lynnagara replied:
Thanks for the explanation about the parens. I'm not sure I follow why `a + b` would be resolved as function names though since function name is not allowed inside arithmetic expression. Only function call is from what I can see, and that wouldn't match the same pattern.

lynnagara replied:
About the date it was just an example of a string that might have a arithmetic expression in it. You can test any string that contains the "+" or "*" characters.

ayirr7 replied:
I think the reason why "a + b" would be resolved as function names is due to the fact that function_name is in the root_element. Hence, the parser would see "a" first and that does match with function_name grammar rule.

ayirr7 replied:
> About the date it was just an example of a string that might have an arithmetic expression in it. You can test any string that contains the "+" or "*" characters in the middle.

This is another reason why the top-level parentheses are being used. "2020-06-06" will not be recognized as an arithmetic expression unless we input it like: "(2020-06-06)."

ayirr7 replied:
Also, string literals were not part of the root_element in the original parser and still are not. So if we were to input "2020-06-06," it would be resolved as a function_name.

fpacifici replied:
Actually @lynnagara made me notice one thing that we can fix: `(column_name)` now would be a valid aggregation. It should not, so adding full arithmetic expressions at the top level inn parentheses has a problem.

What about this, that would simplify the grammar and make it easier to port to snql:
1) remove `function_name` entirely from the grammar at the top. Now at the top you have only arithmetic expression.
2) deal with the function name before line 152. just check if the string you are about to provide to the parser can actually be a function name. (just apply the same regex with the `re` module. `[a-zA-Z_][a-zA-Z0-9_]*`
3) if it is a valid function name, then treat it as such and go what happens at line 206.
4) if not, then apply the parser as you do today. Now every arithmetic expression including simple terms will be valid.

What do you think ?

ayirr7 replied:
Sounds good. I simplified the grammar and some of the visitor code. Let me know what you think!

lynnagara replied:
I think that looks a lot cleaner, thanks

ayirr7 committed:
Merge branch 'master' into feat/arithmeticExpParser

fpacifici commented on snuba/query/parser/expressions.py:
```diff
@@ -21,22 +21,26 @@
+close_paren           = ")"
+space                 = " "
+comma                 = ","
+empty                 = ''
```
please use double quotes

fpacifici commented on snuba/query/parser/expressions.py:
```diff
@@ -52,6 +56,44 @@ def visit_function_name(self, node: Node, visited_children: Iterable[Any]) -> st
+        if exp is None:
+            return term
+        else:
+            return plus(term, exp[1])
```
Do you expect to use `exp[1]` becasue `exp[0]` is the "+" string ?
If yes, mind adding a quick comment above to explain this? Also the type inn the signature of the method (`List[Expression]`) is wrong. It is more likely going to be `Tuple[str, Expression]`

ayirr7 replied:
Added comment and modified type hint.

fpacifici replied:
I think the point of this comment is to explain why it is not exp[0] that you are returning. So explain that exp[0] is the string that represent the operator.

fpacifici commented on snuba/query/parser/expressions.py:
```diff
@@ -52,6 +56,44 @@ def visit_function_name(self, node: Node, visited_children: Iterable[Any]) -> st
+            # that arose from empty space ''
+            return factor
+        else:
+            return multiply(factor, term[1])
```
Same concern I had in the previous comment.

fpacifici reviewed (commented):
Looks good to me. Please just add the test @lynnagara suggested and check the comments inline then I will accept if Lyn is ok with the explanation for the parentheses.

ayirr7 committed:
Added test case - string literal that looks like arithmetic expression.

ayirr7 committed:
Simplification of grammar by removing function_name and function_call from root_element.
Included a test for function_names in order to determine whether or not to
use as input to parser.

ayirr7 committed:
Merge branch 'master' into feat/arithmeticExpParser

fpacifici commented on snuba/query/parser/expressions.py:
```diff
@@ -21,22 +22,27 @@
+
+root_element          = low_pri_arithmetic
+expression            = low_pri_arithmetic / function_call / simple_term
+top_arithm_expression = open_paren low_pri_arithmetic close_paren
```
This rule is not needed anymore, please remove.

ayirr7 committed:
Removed old code (top_arithm_expression grammar rule and visitor function).

ayirr7 committed:
Merge branch 'master' into feat/arithmeticExpParser

fpacifici commented on snuba/query/parser/expressions.py:
```diff
@@ -52,6 +58,50 @@ def visit_function_name(self, node: Node, visited_children: Iterable[Any]) -> st
+    def visit_top_arithm_expression(
+        self, node: Node, visited_children: Tuple[Any, Expression, Any]
+    ) -> Expression:
+        _, exp, _ = visited_children
```
not needed anymore.

fpacifici commented on snuba/query/parser/expressions.py:
```diff
@@ -159,19 +207,29 @@ def parse_aggregation(
-        return CurriedFunctionCall(alias, parsed_expression, tuple(columns_expr),)
+    matched = re.fullmatch("[a-zA-Z_][a-zA-Z0-9_]*", aggregation_function)
+
+    if not bool(matched):
```
you can make this code more readable by inverting the condition (reduces the indentation).

```
if matched is not None:
   return FunctionCall(alias, aggregation_function, tuple(columns_expr))

elif ....
 ....
```

fpacifici commented on tests/query/parser/test_expressions.py:
```diff
@@ -9,6 +9,117 @@
 
 test_data = [
+    (
+        ["f('2020/07/16')", None, None],
```
your arithmetic expression parser does not deal with division anyway :D yet.
Please turn it into an actually possible arithmetic expression: `'2020-7-16'`

ayirr7 replied:
I replaced "/" with the symbols currently supported in arithmetic: "+" and "*".

fpacifici commented on snuba/query/parser/expressions.py:
```diff
@@ -21,22 +22,26 @@
-space           = " "
-comma           = ","
+
+root_element          = low_pri_arithmetic
```
I think these names are a bit misleading, expecially the difference between expression and the term in an arithmetic operation:
why not:
```
root_element = low_pri_arithmetic
low_pri_arithmetic = ... as it is today
high_pri_arithemtic = space* / arithmetic_term / ....the rest is how it is today
arithmetic_term = function_call / numeric_literal / column_name
...
param_expression = low_pri_arithmetic / quoted_literal
parameters_list       = parameter* (param_expression)
parameter             = param_expression space* comma space*
....
```

ayirr7 replied:
Done.

fpacifici reviewed (commented):
We are almost there, thanks for making the changes. Just a couple of cosmetics.
Also @lynnagara are you ok with this version ?

lynnagara commented on snuba/query/parser/expressions.py:
```diff
@@ -159,19 +199,29 @@ def parse_aggregation(
-        # Mix of clickhouse syntax and snuba syntax that generates a CurriedFunction
-        # ["f(a)", "b", None]
-        return CurriedFunctionCall(alias, parsed_expression, tuple(columns_expr),)
+    matched = re.fullmatch("[a-zA-Z_][a-zA-Z0-9_]*", aggregation_function)
```
Let's not duplicate this regex if possible since it's the same one as function name. Can we reference it from minimal_clickhouse_grammar instead?

ayirr7 replied:
To avoid duplicates, we declared the function_name regex outside of the grammar, then used it in the grammar and in the regex matcher. Let me know what you think!

ayirr7 committed:
Modified the test case from "/" to current supported signs ("+", "*").

ayirr7 committed:
Removed duplicate regex in function_name check, simplified grammar, revised comments.

fpacifici commented on snuba/query/parser/expressions.py:
```diff
@@ -159,19 +205,29 @@ def parse_aggregation(
+    if matched is not None:
+        return FunctionCall(alias, aggregation_function, tuple(columns_expr))
+
     else:
```
Since the previous `if` has only one statement inside, you can remove the else block:
```
if matched is not None:
    return FuncitonCall...

expression_tree = minimal_clickhouse_grammar.parse(aggregation_function)
.. all the rest ...

fpacifici commented on snuba/query/parser/expressions.py:
```diff
@@ -159,19 +205,29 @@ def parse_aggregation(
-        # Mix of clickhouse syntax and snuba syntax that generates a CurriedFunction
-        # ["f(a)", "b", None]
-        return CurriedFunctionCall(alias, parsed_expression, tuple(columns_expr),)
+    matched = re.fullmatch(function_name_regex, aggregation_function)
```
please compile the regex before using outside the method
```
FUNCTION_CALL_RE = re.compile(function_name_regex)
```

fpacifici reviewed (approved):
This is great thanks, two comments inline to address before merging but it is good.

ayirr7 committed:
Fixed regex check such that regex compilation happens only once.

ayirr7 committed:
Merge branch 'master' into feat/arithmeticExpParser

ayirr7 committed:
Shortened function_name check code.

fpacifici commented on snuba/query/parser/expressions.py:
```diff
@@ -16,27 +17,36 @@
 from snuba.util import is_function
 
+
+function_name_regex = r"[a-zA-Z_][a-zA-Z0-9_]*"
```
This should be a constant as well, please rename it to `FUNCTION_NAME_REGEX`

ayirr7 committed:
Cosmetic changes.

ayirr7 committed:
Renamed function_name_regex to be a constant.

ayirr7 committed:
Merge branch 'master' into feat/arithmeticExpParser
